File: NyaLauncher.Core/Download/ModLoaderMetadata.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace NyaLauncher.Core.Download;

/// <summary>
/// 支持的 Mod Loader 类型。
/// </summary>
public enum ModLoaderType
{
    Vanilla,
    Fabric,
    NeoForge,
    Forge
}

/// <summary>
/// 单个 Mod Loader 版本条目（从 Loader API 获取）。
/// </summary>
public sealed record ModLoaderVersion
{
    /// <summary>Loader 类型。</summary>
    public ModLoaderType Type { get; init; }

    /// <summary>Loader 版本号，如 "0.16.14"（Fabric）或 "21.8.1"（NeoForge）。</summary>
    public string LoaderVersion { get; init; } = string.Empty;

    /// <summary>是否为推荐/稳定版本。</summary>
    public bool IsStable { get; init; }

    /// <summary>构建号（部分 Loader 使用）。</summary>
    public int BuildNumber { get; init; }

    /// <summary>安装此 Loader 所需的版本 JSON 元数据 URL。</summary>
    public string MetadataUrl { get; init; } = string.Empty;

    /// <summary>
    /// MetadataUrl 指向的是安装器 JAR 而非版本 JSON。
    /// 需要下载 JAR 后从中提取 version.json 或 install_profile.json。
    /// </summary>
    public bool RequiresInstallerExtraction { get; init; }

    /// <summary>用于 UI 展示的格式化名称。</summary>
    public string DisplayName => Type switch
    {
        ModLoaderType.Fabric => $"Fabric Loader {LoaderVersion}",
        ModLoaderType.NeoForge => $"NeoForge {LoaderVersion}",
        ModLoaderType.Forge => $"Forge {LoaderVersion}",
        _ => LoaderVersion
    };
}

/// <summary>
/// Mod Loader 元数据获取服务。从各 Loader 官方 API 获取可用版本列表和安装元数据。
/// </summary>
public static class ModLoaderMetadata
{
    // ------------------------------------------------------------------
    // Fabric
    // ------------------------------------------------------------------

    private const string FabricLoaderVersionsUrl =
        "https://meta.fabricmc.net/v2/versions/loader/";

    /// <summary>
    /// 获取指定 Minecraft 版本可用的 Fabric Loader 版本列表。
    /// </summary>
    public static async Task<List<ModLoaderVersion>> GetFabricVersionsAsync(
        string minecraftVersion,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(minecraftVersion);

        var url = $"{FabricLoaderVersionsUrl}{minecraftVersion}";
        var json = await DownloadSourceProvider.GetStringAsync(url, TimeSpan.FromSeconds(10), cancellationToken)
            .ConfigureAwait(false);
        var response = System.Text.Json.JsonSerializer.Deserialize<List<FabricLoaderEntry>>(json);

        if (response is null)
            return [];

        return response
            .Where(entry => entry.Loader is not null)
            .Select(entry => new ModLoaderVersion
            {
                Type = ModLoaderType.Fabric,
                LoaderVersion = entry.Loader!.Version,
                IsStable = entry.Loader.Stable,
                MetadataUrl = $"{FabricLoaderVersionsUrl}{minecraftVersion}/{entry.Loader.Version}/profile/json"
            })
            .ToList();
    }

    // ------------------------------------------------------------------
    // NeoForge
    // ------------------------------------------------------------------

    private const string NeoForgeVersionsUrl =
        "https://maven.neoforged.net/releases/net/neoforged/neoforge/maven-metadata.xml";

    /// <summary>BMCLAPI（bangbang93 镜像）按 Minecraft 版本查询 NeoForge 列表。</summary>
    private const string NeoForgeBmclListUrl =
        "https://bmclapi2.bangbang93.com/neoforge/list/{0}";

    /// <summary>BMCLAPI 下载 NeoForge 安装器 JAR（302 重定向到实际文件）。</summary>
    private const string NeoForgeBmclInstallerUrl =
        "https://bmclapi2.bangbang93.com/neoforge/version/{0}/download/installer.jar";

    /// <summary>
    /// 获取指定 Minecraft 版本可用的 NeoForge 版本列表。
    /// 优先使用 BMCLAPI 镜像（国内网络可达性好），失败时回退官方 Maven。
    /// NeoForge 版本格式为 "{mcMajor}.{mcMinor}.{patch}"，如 "21.8.1"。
    /// </summary>
    public static async Task<List<ModLoaderVersion>> GetNeoForgeVersionsAsync(
        string minecraftVersion,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(minecraftVersion);

        // 1) 优先 BMCLAPI：直接返回该 MC 版本下的 NeoForge 版本，无需前缀推导
        try
        {
            var bmclJson = await DownloadSourceProvider.GetStringAsync(
                    string.Format(NeoForgeBmclListUrl, minecraftVersion),
                    TimeSpan.FromSeconds(10), cancellationToken)
                .ConfigureAwait(false);
            var entries = JsonSerializer.Deserialize<List<NeoForgeBmclEntry>>(bmclJson);

            if (entries is { Count: > 0 })
            {
                return entries
                    .Where(e => !string.IsNullOrWhiteSpace(e.Version))
                    .Select(e => NormalizeBmclNeoForgeVersion(e.Version))
                    .Where(v => !string.IsNullOrWhiteSpace(v))
                    .Distinct(StringComparer.Ordinal)
                    .OrderByDescending(v => v, Comparer<string>.Create(ModrinthVersionApi.CompareVersionStrings))
                    .Select(v => new ModLoaderVersion
                    {
                        Type = ModLoaderType.NeoForge,
                        LoaderVersion = v,
                        MetadataUrl = string.Format(NeoForgeBmclInstallerUrl, v),
                        RequiresInstallerExtraction = true
                    })
                    .ToList();
            }
        }
        catch
        {
            // BMCL 不可用（网络/格式异常）时回退官方源，避免版本列表为空
        }

        // 2) 回退：官方 maven-metadata.xml（按前缀匹配）
        var xml = await DownloadSourceProvider.GetStringAsync(
                NeoForgeVersionsUrl, TimeSpan.FromSeconds(10), cancellationToken)
            .ConfigureAwait(false);

        var versions = ParseMavenVersions(xml);
        var prefix = ToNeoForgePrefix(minecraftVersion);

        return versions
            // 带版本边界匹配：仅匹配 "21.8.x" 或 "21.0.x"（基础版 1.21 → 21.0.x），
            // 防止 "21.1" 误匹配 "21.11.x"、防止 "21" 误匹配全部 21.x 系列。
            .Where(v => MatchesPrefix(v, prefix))
            // 按版本号数字比较降序（字典序会让 21.8.9 排在 21.8.54 前面）
            .OrderByDescending(v => v, Comparer<string>.Create(ModrinthVersionApi.CompareVersionStrings))
            .Select(v => new ModLoaderVersion
            {
                Type = ModLoaderType.NeoForge,
                LoaderVersion = v,
                MetadataUrl =
                    $"https://maven.neoforged.net/releases/net/neoforged/neoforge/{v}/neoforge-{v}-installer.jar",
                RequiresInstallerExtraction = true
            })
            .ToList();
    }

    /// <summary>
    /// 归一化 BMCL 返回的 NeoForge 版本号：
    /// 纯版本号（如 "21.8.54"）原样返回；
    /// 带 "mc-版本" 前缀的（如 "1.20.1-47.1.12"）截取纯版本号。
    /// </summary>
    internal static string NormalizeBmclNeoForgeVersion(string version)
    {
        var v = version.Trim();
        if (v.Length == 0 || !char.IsDigit(v[0]))
            return v;

        var dash = v.IndexOf('-');
        if (dash > 0 && dash < v.Length - 1)
        {
            var tail = v[(dash + 1)..];
            if (char.IsDigit(tail[0]))
                return tail;
        }
        return v;
    }

    // ------------------------------------------------------------------
    // Forge
    // ------------------------------------------------------------------

    private const string ForgePromotionsUrl =
        "https://files.minecraftforge.net/net/minecraftforge/forge/promotions_slim.json";

    /// <summary>
    /// 获取指定 Minecraft 版本推荐的 Forge 版本。
    /// Forge 的版本分发方式不同于 Fabric/NeoForge，只提供推荐版本。
    /// </summary>
    public static async Task<List<ModLoaderVersion>> GetForgeVersionsAsync(
        string minecraftVersion,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(minecraftVersion);

        var json = await DownloadSourceProvider.GetStringAsync(
                ForgePromotionsUrl, TimeSpan.FromSeconds(10), cancellationToken)
            .ConfigureAwait(false);

        using var doc = System.Text.Json.JsonDocument.Parse(json);
        if (!doc.RootElement.TryGetProperty("promos", out var promos))
            return [];

        var results = new List<ModLoaderVersion>();
        foreach (var property in promos.EnumerateObject())
        {
            // key 格式: "{mcVersion}-latest" 或 "{mcVersion}-recommended"
            // 必须与所选 MC 版本精确匹配：防止 "1.21" 误匹配 "1.21.1-latest"
            if (!MatchesForgePromoKey(property.Name, minecraftVersion))
                continue;

            var forgeVersion = property.Value.GetString();
            if (string.IsNullOrWhiteSpace(forgeVersion))
                continue;

            var isRecommended = property.Name.EndsWith("-recommended", StringComparison.OrdinalIgnoreCase);
            var fullVersion = $"{minecraftVersion}-{forgeVersion}";

            results.Add(new ModLoaderVersion
            {
                Type = ModLoaderType.Forge,
                LoaderVersion = fullVersion,
                IsStable = isRecommended,
                MetadataUrl =
                    $"https://maven.minecraftforge.net/net/minecraftforge/forge/{fullVersion}/forge-{fullVersion}-installer.jar",
                RequiresInstallerExtraction = true
            });
        }

        return results
            .OrderByDescending(v => v.IsStable)
            .ThenByDescending(v => v.LoaderVersion)
            .ToList();
    }

    // ------------------------------------------------------------------
    // 通用入口
    // ------------------------------------------------------------------

    /// <summary>
    /// 获取指定 Loader 类型在指定 Minecraft 版本下的可用版本列表。
    /// </summary>
    public static Task<List<ModLoaderVersion>> GetVersionsAsync(
        ModLoaderType type,
        string minecraftVersion,
        CancellationToken cancellationToken = default) => type switch
    {
        ModLoaderType.Fabric => GetFabricVersionsAsync(minecraftVersion, cancellationToken),
        ModLoaderType.NeoForge => GetNeoForgeVersionsAsync(minecraftVersion, cancellationToken),
        ModLoaderType.Forge => GetForgeVersionsAsync(minecraftVersion, cancellationToken),
        _ => Task.FromResult<List<ModLoaderVersion>>([])
    };

    // ------------------------------------------------------------------
    // 辅助方法
    // ------------------------------------------------------------------

    /// <summary>
    /// 将 Minecraft 版本号转换为 NeoForge 的 Maven 版本前缀。
    /// 如 "1.21.8" → "21.8"，"1.21"（基础版）→ "21.0"。
    /// </summary>
    internal static string ToNeoForgePrefix(string minecraftVersion)
    {
        // NeoForge 版本格式: {mcMinor}.{mcPatch}
        // MC 1.21.8 → NeoForge 21.8.x；MC 1.21（无补丁号）→ NeoForge 21.0.x
        var parts = minecraftVersion.Split('.');
        if (parts.Length >= 3)
            return $"{parts[1]}.{parts[2]}";
        if (parts.Length == 2)
            return $"{parts[1]}.0";
        return minecraftVersion;
    }

    /// <summary>
    /// 带版本边界的精确前缀匹配：仅当 value 等于 prefix 或以 "prefix." 开头时匹配，
    /// 防止 "21.1" 误匹配 "21.11.x" 这类错误。
    /// </summary>
    internal static bool MatchesPrefix(string value, string prefix) =>
        value.Equals(prefix, StringComparison.OrdinalIgnoreCase) ||
        value.StartsWith(prefix + ".", StringComparison.OrdinalIgnoreCase);

    /// <summary>
    /// Forge promos 键匹配：键格式为 "{mcVersion}-latest" / "{mcVersion}-recommended"，
    /// 要求去后缀后与所选 MC 版本精确相等，防止 "1.21" 误匹配 "1.21.1-latest"。
    /// </summary>
    internal static bool MatchesForgePromoKey(string promoKey, string minecraftVersion)
    {
        if (string.IsNullOrWhiteSpace(promoKey) || string.IsNullOrWhiteSpace(minecraftVersion))
            return false;
        foreach (var suffix in new[] { "-latest", "-recommended" })
        {
            if (!promoKey.EndsWith(suffix, StringComparison.OrdinalIgnoreCase))
                continue;
            return string.Equals(
                promoKey[..^suffix.Length],
                minecraftVersion,
                StringComparison.OrdinalIgnoreCase);
        }
        return false;
    }

    /// <summary>
    /// 解析 Maven maven-metadata.xml 中的 version 列表。
    /// </summary>
    private static List<string> ParseMavenVersions(string xml)
    {
        var versions = new List<string>();
        const string startTag = "<version>";
        const string endTag = "</version>";

        var span = xml.AsSpan();
        while (true)
        {
            var startIndex = span.IndexOf(startTag.AsSpan());
            if (startIndex < 0)
                break;

            var valueStart = startIndex + startTag.Length;
            var endIndex = span[valueStart..].IndexOf(endTag.AsSpan());
            if (endIndex < 0)
                break;

            versions.Add(span.Slice(valueStart, endIndex).ToString());
            span = span[(valueStart + endIndex + endTag.Length)..];
        }

        return versions;
    }

    // ------------------------------------------------------------------
    // JSON 反序列化模型
    // ------------------------------------------------------------------

    private sealed class FabricLoaderEntry
    {
        [JsonPropertyName("loader")]
        public FabricLoaderInfo? Loader { get; set; }
    }

    private sealed class FabricLoaderInfo
    {
        [JsonPropertyName("version")]
        public string Version { get; set; } = string.Empty;

        [JsonPropertyName("stable")]
        public bool Stable { get; set; }
    }

    /// <summary>BMCLAPI /neoforge/list/:mcversion 返回条目。</summary>
    private sealed class NeoForgeBmclEntry
    {
        [JsonPropertyName("rawVersion")]
        public string RawVersion { get; set; } = string.Empty;

        [JsonPropertyName("version")]
        public string Version { get; set; } = string.Empty;

        [JsonPropertyName("mcversion")]
        public string McVersion { get; set; } = string.Empty;
    }
}

File: internal/download/mod_loader_metadata.go
package download

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"

	"nyalauncher/internal/download/modrinth"
)

// ModLoaderType 支持的 Mod Loader 类型。
type ModLoaderType int

const (
	ModLoaderVanilla ModLoaderType = iota
	ModLoaderFabric
	ModLoaderQuilt
	ModLoaderNeoForge
	ModLoaderForge
)

// String 枚举名（与 C# 枚举名一致，用于持久化与显示）。
func (t ModLoaderType) String() string {
	switch t {
	case ModLoaderFabric:
		return "Fabric"
	case ModLoaderQuilt:
		return "Quilt"
	case ModLoaderNeoForge:
		return "NeoForge"
	case ModLoaderForge:
		return "Forge"
	default:
		return "Vanilla"
	}
}

// ParseModLoaderType 从名称解析 Loader 类型（大小写不敏感）。
func ParseModLoaderType(name string) (ModLoaderType, bool) {
	switch strings.ToLower(strings.TrimSpace(name)) {
	case "vanilla":
		return ModLoaderVanilla, true
	case "fabric":
		return ModLoaderFabric, true
	case "quilt":
		return ModLoaderQuilt, true
	case "neoforge":
		return ModLoaderNeoForge, true
	case "forge":
		return ModLoaderForge, true
	default:
		return ModLoaderVanilla, false
	}
}

// ModLoaderVersion 单个 Mod Loader 版本条目（从 Loader API 获取）。
type ModLoaderVersion struct {
	// Type Loader 类型。
	Type ModLoaderType
	// LoaderVersion Loader 版本号，如 "0.16.14"（Fabric）或 "21.8.1"（NeoForge）。
	LoaderVersion string
	// IsStable 是否为推荐/稳定版本。
	IsStable bool
	// BuildNumber 构建号（部分 Loader 使用）。
	BuildNumber int
	// MetadataURL 安装此 Loader 所需的版本 JSON 元数据 URL。
	MetadataURL string
	// RequiresInstallerExtraction MetadataURL 指向的是安装器 JAR 而非版本 JSON。
	// 需要下载 JAR 后从中提取 version.json 或 install_profile.json。
	RequiresInstallerExtraction bool
}

// DisplayName 用于 UI 展示的格式化名称。
func (v ModLoaderVersion) DisplayName() string {
	switch v.Type {
	case ModLoaderFabric:
		return fmt.Sprintf("Fabric Loader %s", v.LoaderVersion)
	case ModLoaderQuilt:
		return fmt.Sprintf("Quilt Loader %s", v.LoaderVersion)
	case ModLoaderNeoForge:
		return fmt.Sprintf("NeoForge %s", v.LoaderVersion)
	case ModLoaderForge:
		return fmt.Sprintf("Forge %s", v.LoaderVersion)
	default:
		return v.LoaderVersion
	}
}

// ModLoaderMetadata Mod Loader 元数据获取服务。
// 从各 Loader 官方 API 获取可用版本列表和安装元数据。对应 C# ModLoaderMetadata。

const fabricLoaderVersionsURL = "https://meta.fabricmc.net/v2/versions/loader/"
const quiltLoaderVersionsURL = "https://meta.quiltmc.org/v3/versions/loader/"

const neoForgeVersionsURL = "https://maven.neoforged.net/releases/net/neoforged/neoforge/maven-metadata.xml"

// neoForgeBmclListURL BMCLAPI（bangbang93 镜像）按 Minecraft 版本查询 NeoForge 列表。
const neoForgeBmclListURL = "https://bmclapi2.bangbang93.com/neoforge/list/%s"

// neoForgeBmclInstallerURL BMCLAPI 下载 NeoForge 安装器 JAR（302 重定向到实际文件）。
const neoForgeBmclInstallerURL = "https://bmclapi2.bangbang93.com/neoforge/version/%s/download/installer.jar"

const forgePromotionsURL = "https://files.minecraftforge.net/net/minecraftforge/forge/promotions_slim.json"

// fabricLoaderEntry Fabric/Quilt meta API 响应条目。
type fabricLoaderEntry struct {
	Loader *fabricLoaderInfo `json:"loader"`
}

type fabricLoaderInfo struct {
	Version string `json:"version"`
	Stable  bool   `json:"stable"`
}

// neoForgeBmclEntry BMCLAPI /neoforge/list/:mcversion 返回条目。
type neoForgeBmclEntry struct {
	RawVersion string `json:"rawVersion"`
	Version    string `json:"version"`
	McVersion  string `json:"mcversion"`
}

// loaderMetadataTimeout Loader 元数据请求的 10 秒超时（与 C# 各调用点一致）。
const loaderMetadataTimeout = 10 * time.Second

func shortTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
	return context.WithTimeout(ctx, loaderMetadataTimeout)
}

// GetFabricVersions 获取指定 Minecraft 版本可用的 Fabric Loader 版本列表。
func GetFabricVersions(ctx context.Context, minecraftVersion string) ([]ModLoaderVersion, error) {
	if strings.TrimSpace(minecraftVersion) == "" {
		return nil, fmt.Errorf("minecraftVersion 不能为空")
	}

	ctx, cancel := shortTimeout(ctx)
	defer cancel()
	rawJSON, err := SourceProvider.GetString(ctx, fabricLoaderVersionsURL+minecraftVersion, nil)
	if err != nil {
		return nil, err
	}
	var response []fabricLoaderEntry
	if err := json.Unmarshal([]byte(rawJSON), &response); err != nil {
		return nil, err
	}

	result := make([]ModLoaderVersion, 0, len(response))
	for _, entry := range response {
		if entry.Loader == nil {
			continue
		}
		result = append(result, ModLoaderVersion{
			Type:          ModLoaderFabric,
			LoaderVersion: entry.Loader.Version,
			IsStable:      entry.Loader.Stable,
			MetadataURL: fmt.Sprintf("%s%s/%s/profile/json",
				fabricLoaderVersionsURL, minecraftVersion, entry.Loader.Version),
		})
	}
	return result, nil
}

// GetQuiltVersions 获取指定 Minecraft 版本可用的 Quilt Loader 版本列表。
// Quilt 的 meta API（v3）响应形状与 Fabric v2 相同，可直接复用 Fabric 的反序列化模型；
// profile JSON 同样是标准版本 JSON（inheritsFrom + Knot 启动主类），
// 库文件来自 maven.quiltmc.org 与 maven.fabricmc.org。
func GetQuiltVersions(ctx context.Context, minecraftVersion string) ([]ModLoaderVersion, error) {
	if strings.TrimSpace(minecraftVersion) == "" {
		return nil, fmt.Errorf("minecraftVersion 不能为空")
	}

	ctx, cancel := shortTimeout(ctx)
	defer cancel()
	rawJSON, err := SourceProvider.GetString(ctx, quiltLoaderVersionsURL+minecraftVersion, nil)
	if err != nil {
		return nil, err
	}
	var response []fabricLoaderEntry
	if err := json.Unmarshal([]byte(rawJSON), &response); err != nil {
		return nil, err
	}

	result := make([]ModLoaderVersion, 0, len(response))
	for _, entry := range response {
		if entry.Loader == nil {
			continue
		}
		result = append(result, ModLoaderVersion{
			Type:          ModLoaderQuilt,
			LoaderVersion: entry.Loader.Version,
			IsStable:      entry.Loader.Stable,
			MetadataURL: fmt.Sprintf("%s%s/%s/profile/json",
				quiltLoaderVersionsURL, minecraftVersion, entry.Loader.Version),
		})
	}
	return result, nil
}

// GetNeoForgeVersions 获取指定 Minecraft 版本可用的 NeoForge 版本列表。
// 按用户选择的下载源决定优先顺序：BMCL 源（默认）优先走 BMCLAPI 镜像
// （国内网络可达性好），Official 源优先走官方 Maven，另一侧仅作兜底。
// NeoForge 版本格式为 "{mcMajor}.{mcMinor}.{patch}"，如 "21.8.1"。
func GetNeoForgeVersions(ctx context.Context, minecraftVersion string) ([]ModLoaderVersion, error) {
	if strings.TrimSpace(minecraftVersion) == "" {
		return nil, fmt.Errorf("minecraftVersion 不能为空")
	}

	if SourceProvider.Active().Name != DownloadSources.Official.Name {
		if fromBmcl := tryFetchNeoForgeFromBmcl(ctx, minecraftVersion); fromBmcl != nil {
			return fromBmcl, nil
		}
		return fetchNeoForgeFromOfficialMaven(ctx, minecraftVersion)
	}

	fromOfficial, err := fetchNeoForgeFromOfficialMaven(ctx, minecraftVersion)
	if err == nil && len(fromOfficial) > 0 {
		return fromOfficial, nil
	}
	// 官方源不可用（网络/格式异常）时兜底 BMCL，避免版本列表为空
	if fromBmcl := tryFetchNeoForgeFromBmcl(ctx, minecraftVersion); fromBmcl != nil {
		return fromBmcl, nil
	}
	return []ModLoaderVersion{}, nil
}

// tryFetchNeoForgeFromBmcl BMCLAPI（bangbang93 镜像）获取 NeoForge 版本列表：
// 直接返回该 MC 版本下的版本，无需前缀推导；不可用或无结果返回 nil。
func tryFetchNeoForgeFromBmcl(ctx context.Context, minecraftVersion string) []ModLoaderVersion {
	ctx, cancel := shortTimeout(ctx)
	defer cancel()
	bmclJSON, err := SourceProvider.GetString(ctx, fmt.Sprintf(neoForgeBmclListURL, minecraftVersion), nil)
	if err != nil {
		return nil
	}
	var entries []neoForgeBmclEntry
	if err := json.Unmarshal([]byte(bmclJSON), &entries); err != nil {
		return nil
	}

	seen := map[string]bool{}
	var versions []string
	for _, e := range entries {
		if strings.TrimSpace(e.Version) == "" {
			continue
		}
		v := NormalizeBmclNeoForgeVersion(e.Version)
		if v == "" || seen[v] {
			continue
		}
		seen[v] = true
		versions = append(versions, v)
	}
	sort.SliceStable(versions, func(i, j int) bool {
		return modrinth.CompareVersionStrings(versions[i], versions[j]) > 0
	})

	result := make([]ModLoaderVersion, 0, len(versions))
	for _, v := range versions {
		result = append(result, ModLoaderVersion{
			Type:                        ModLoaderNeoForge,
			LoaderVersion:               v,
			MetadataURL:                 fmt.Sprintf(neoForgeBmclInstallerURL, v),
			RequiresInstallerExtraction: true,
		})
	}
	return result
}

// fetchNeoForgeFromOfficialMaven 官方 Maven maven-metadata.xml 获取 NeoForge 版本列表（按前缀匹配）。
func fetchNeoForgeFromOfficialMaven(ctx context.Context, minecraftVersion string) ([]ModLoaderVersion, error) {
	ctx, cancel := shortTimeout(ctx)
	defer cancel()
	metadataXML, err := SourceProvider.GetString(ctx, neoForgeVersionsURL, nil)
	if err != nil {
		return nil, err
	}

	versions := parseMavenVersions(metadataXML)
	prefix := ToNeoForgePrefix(minecraftVersion)

	matched := make([]string, 0, len(versions))
	for _, v := range versions {
		// 带版本边界匹配：仅匹配 "21.8.x" 或 "21.0.x"（基础版 1.21 → 21.0.x），
		// 防止 "21.1" 误匹配 "21.11.x"、防止 "21" 误匹配全部 21.x 系列。
		if matchesPrefix(v, prefix) {
			matched = append(matched, v)
		}
	}
	// 按版本号数字比较降序（字典序会让 21.8.9 排在 21.8.54 前面）
	sort.SliceStable(matched, func(i, j int) bool {
		return modrinth.CompareVersionStrings(matched[i], matched[j]) > 0
	})

	result := make([]ModLoaderVersion, 0, len(matched))
	for _, v := range matched {
		result = append(result, ModLoaderVersion{
			Type:          ModLoaderNeoForge,
			LoaderVersion: v,
			MetadataURL: fmt.Sprintf(
				"https://maven.neoforged.net/releases/net/neoforged/neoforge/%s/neoforge-%s-installer.jar", v, v),
			RequiresInstallerExtraction: true,
		})
	}
	return result, nil
}

// NormalizeBmclNeoForgeVersion 归一化 BMCL 返回的 NeoForge 版本号：
// 纯版本号（如 "21.8.54"）原样返回；带 "mc-版本" 前缀的（如 "1.20.1-47.1.12"）截取纯版本号。
func NormalizeBmclNeoForgeVersion(version string) string {
	v := strings.TrimSpace(version)
	if len(v) == 0 || v[0] < '0' || v[0] > '9' {
		return v
	}
	dash := strings.Index(v, "-")
	if dash > 0 && dash < len(v)-1 {
		tail := v[dash+1:]
		if tail[0] >= '0' && tail[0] <= '9' {
			return tail
		}
	}
	return v
}

// GetForgeVersions 获取指定 Minecraft 版本推荐的 Forge 版本。
// Forge 的版本分发方式不同于 Fabric/NeoForge，只提供推荐版本。
func GetForgeVersions(ctx context.Context, minecraftVersion string) ([]ModLoaderVersion, error) {
	if strings.TrimSpace(minecraftVersion) == "" {
		return nil, fmt.Errorf("minecraftVersion 不能为空")
	}

	ctx, cancel := shortTimeout(ctx)
	defer cancel()
	rawJSON, err := SourceProvider.GetString(ctx, forgePromotionsURL, nil)
	if err != nil {
		return nil, err
	}

	var doc struct {
		Promos map[string]string `json:"promos"`
	}
	if err := json.Unmarshal([]byte(rawJSON), &doc); err != nil {
		return nil, err
	}

	var results []ModLoaderVersion
	for key, forgeVersion := range doc.Promos {
		// key 格式: "{mcVersion}-latest" 或 "{mcVersion}-recommended"
		// 必须与所选 MC 版本精确匹配：防止 "1.21" 误匹配 "1.21.1-latest"
		if !matchesForgePromoKey(key, minecraftVersion) {
			continue
		}
		if strings.TrimSpace(forgeVersion) == "" {
			continue
		}

		isRecommended := strings.HasSuffix(strings.ToLower(key), "-recommended")
		fullVersion := fmt.Sprintf("%s-%s", minecraftVersion, forgeVersion)

		results = append(results, ModLoaderVersion{
			Type:          ModLoaderForge,
			LoaderVersion: fullVersion,
			IsStable:      isRecommended,
			MetadataURL: fmt.Sprintf(
				"https://maven.minecraftforge.net/net/minecraftforge/forge/%s/forge-%s-installer.jar",
				fullVersion, fullVersion),
			RequiresInstallerExtraction: true,
		})
	}

	sort.SliceStable(results, func(i, j int) bool {
		if results[i].IsStable != results[j].IsStable {
			return results[i].IsStable
		}
		return results[i].LoaderVersion > results[j].LoaderVersion
	})
	return results, nil
}

// GetModLoaderVersions 获取指定 Loader 类型在指定 Minecraft 版本下的可用版本列表。
func GetModLoaderVersions(ctx context.Context, loaderType ModLoaderType, minecraftVersion string) ([]ModLoaderVersion, error) {
	switch loaderType {
	case ModLoaderFabric:
		return GetFabricVersions(ctx, minecraftVersion)
	case ModLoaderQuilt:
		return GetQuiltVersions(ctx, minecraftVersion)
	case ModLoaderNeoForge:
		return GetNeoForgeVersions(ctx, minecraftVersion)
	case ModLoaderForge:
		return GetForgeVersions(ctx, minecraftVersion)
	default:
		return []ModLoaderVersion{}, nil
	}
}

// ToNeoForgePrefix 将 Minecraft 版本号转换为 NeoForge 的 Maven 版本前缀。
// 如 "1.21.8" → "21.8"，"1.21"（基础版）→ "21.0"。
func ToNeoForgePrefix(minecraftVersion string) string {
	// NeoForge 版本格式: {mcMinor}.{mcPatch}
	// MC 1.21.8 → NeoForge 21.8.x；MC 1.21（无补丁号）→ NeoForge 21.0.x
	parts := strings.Split(minecraftVersion, ".")
	if len(parts) >= 3 {
		return fmt.Sprintf("%s.%s", parts[1], parts[2])
	}
	if len(parts) == 2 {
		return fmt.Sprintf("%s.0", parts[1])
	}
	return minecraftVersion
}

// matchesPrefix 带版本边界的精确前缀匹配：仅当 value 等于 prefix 或以 "prefix." 开头时匹配，
// 防止 "21.1" 误匹配 "21.11.x" 这类错误。
func matchesPrefix(value, prefix string) bool {
	return strings.EqualFold(value, prefix) ||
		strings.HasPrefix(strings.ToLower(value), strings.ToLower(prefix)+".")
}

// matchesForgePromoKey Forge promos 键匹配：键格式为 "{mcVersion}-latest" / "{mcVersion}-recommended"，
// 要求去后缀后与所选 MC 版本精确相等，防止 "1.21" 误匹配 "1.21.1-latest"。
func matchesForgePromoKey(promoKey, minecraftVersion string) bool {
	if strings.TrimSpace(promoKey) == "" || strings.TrimSpace(minecraftVersion) == "" {
		return false
	}
	for _, suffix := range []string{"-latest", "-recommended"} {
		if !strings.HasSuffix(strings.ToLower(promoKey), suffix) {
			continue
		}
		return strings.EqualFold(promoKey[:len(promoKey)-len(suffix)], minecraftVersion)
	}
	return false
}

// parseMavenVersions 解析 Maven maven-metadata.xml 中的 version 列表。
func parseMavenVersions(metadataXML string) []string {
	var versions []string
	const startTag = "<version>"
	const endTag = "</version>"

	span := metadataXML
	for {
		startIndex := strings.Index(span, startTag)
		if startIndex < 0 {
			break
		}
		valueStart := startIndex + len(startTag)
		rest := span[valueStart:]
		endIndex := strings.Index(rest, endTag)
		if endIndex < 0 {
			break
		}
		versions = append(versions, rest[:endIndex])
		span = rest[endIndex+len(endTag):]
	}
	return versions
}
